PICMI_Python/base.py

```diff
@@ -29,60 +41,131 @@ def register_constants(implementation_constants):
+        
         # "if bases" skips this for the _ClassWithInit (which has no bases)
         # "if bases[0].__doc__ is not None" skips this for the picmistandard classes since their bases[0] (i.e. _ClassWithInit)
         # has no __doc__.
-        if bases and bases[0].__doc__ is not None:
-            implementation_doc = attrs.get('__doc__', '')
+        elif bases and bases[0].__doc__ is not None:
+            implementation_doc = namespace.get('__doc__', '')
             if implementation_doc:
                 # The format of the added string is intentional.
                 # The double return "\n\n" is needed to start a new section in the documentation.
                 # Then the four spaces matches the standard level of indentation for doc strings
                 # (assuming PEP8 formatting).
                 # The final return "\n" assumes that the implementation doc string begins with a return,
                 # i.e. a line with only three quotes, """.
-                attrs['__doc__'] = bases[0].__doc__ + """\n\n    Implementation specific documentation\n""" + implementation_doc
+                namespace['__doc__'] = bases[0].__doc__ + """\n\n    Implementation specific documentation\n""" + implementation_doc
             else:
-                attrs['__doc__'] = bases[0].__doc__
-        return super(_DocumentedMetaClass, cls).__new__(cls, name, bases, attrs)
-
-
-class _ClassWithInit(metaclass=_DocumentedMetaClass):
-    def handle_init(self, kw):
-        # --- Grab all keywords for the current code.
-        # --- Arguments for other supported codes are ignored.
-        # --- If there is anything left over, it is an error.
-        codekw = {}
-        for k,v in kw.copy().items():
-            code = k.split('_')[0]
-            if code == codename:
-                codekw[k] = v
-                kw.pop(k)
-            elif code in supported_codes:
-                kw.pop(k)
-
-        if kw:
-            raise TypeError('Unexpected keyword argument: "%s"'%list(kw))
-
-        # --- It is expected that init strips accepted keywords from codekw.
-        self.init(codekw)
+                namespace['__doc__'] = bases[0].__doc__
+        return super().__new__(mcs, name, bases, namespace, **kwargs)
 
-        if codekw:
-            raise TypeError("Unexpected keyword argument for %s: '%s'"%(codename, list(codekw)))
 
-    def init(self, kw):
-        # --- The implementation of this routine should use kw.pop() to retrieve input arguments from kw.
-        # --- This allows testing for any unused arguments and raising an error if found.
+class _ClassWithInit(BaseModel, metaclass=_DocumentedMetaClass):
+    """
+    Base class for all PICMI classes using Pydantic for validation and extensibility.
+    
+    This class allows code-specific extensions (e.g., warpx_* kwargs) via Pydantic's
+    extra fields mechanism while maintaining type safety for standard attributes.
+    """
```

Comment:
Not sure if all this mechanism is still needed. I would simply derive in downstream code from a PICMI class and extend it with more fields...